token/oauth/oauth2.go

```diff
@@ -81,7 +81,7 @@ func (provider *OauthIdentityProvider) UserProfilePayload(ctx context.Context, t
 }
 
 // SaveReferrer validates referrer and saves it in DB
-func SaveReferrer(ctx context.Context, db application.DB, state string, referrer string, validReferrerURL string) error {
+func SaveReferrer(ctx context.Context, db application.DB, state string, referrer string, responseMode, validReferrerURL string) error {
```

Comment:
"query" and nil were making quite some confusion in implementation. Better approach would be to ignore it if response_mode = "query" (responseMode=nil), it would also save us the trouble of saving responseMode in the database in the next commit including other changes you asked for.